### PR TITLE
feat: read route info from SFC custom block

### PR DESCRIPTION
**Pull request by JohnCampionJr**
Closes #5

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -147,4 +147,30 @@ export default {
...
+name: name-override
+meta:
+  requiresAuth: true
+</route>
+```
+
+
 ### **See more details: [vite-plugin-voie](https://github.com/brattonross/vite-plugin-voie)**
+
+### Route block parsing code: [@ktsn/vue-route-generator](https://github.com/ktsn/vue-route-generator)
```
Comment (hannoeru) [inline README.md:176]: I think we don't need to add this, as we are using the standard way to parse SFC.

**Review by hannoeru — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (hannoeru) [inline README.md:176]: I think we don't need to add this, as we are using the standard way to parse SFC.

File: examples/spa/package.json [hunk 2]
```diff
@@ -1,7 +1,7 @@
 {
   "private": true,
   "scripts": {
-    "dev": "cross-env DEBUG=vite-plugin-pages vite",
+    "dev": "cross-env DEBUG=vite-plugin-pages vite --port 3332",
```
Comment (hannoeru) [inline examples/spa/package.json:4]: Why changed this?

File: examples/spa/src/pages/about/who/me.vue [hunk 3]
```diff
@@ -11,3 +11,8 @@ export default {
 <style>
 
 </style>
+<route>
+name: who-me-override
+meta:
+  requiresAuth: true
+</route>
```
Comment (hannoeru) [inline examples/spa/src/pages/about/who/me.vue:18]: Can you add lang=xxx feature to specify the parser to use?

File: src/parseSfc.ts [hunk 4]
```diff
@@ -26,25 +27,47 @@ export interface FileError extends Error {
...
+    }
+    catch (e) {
+      err = e
+      valid = false
     }
   }
+
+  if (valid) return result
+
+  const wrapped: FileError = new Error(`Invalid JSON/JSON5/YAML format of <${block.type}> content in ${filePath}\n${err.message}`)
```
Comment (hannoeru) [inline src/parseSfc.ts:66]: Use `this.error(/* error msg */)` in L49 & L59

File: src/parseSfc.ts [hunk 5]
```diff
@@ -26,25 +27,47 @@ export interface FileError extends Error {
...
-  content: string,
-  filePath: string,
-  blockName: string): any {
-  try {
-    return JSON5.parse(content)
+export function tryParseCustomBlock(block: CustomBlock, filePath: string): any {
+  const lang = block.lang ?? 'any'
+  let valid = true
+  let err: Error = new Error()
+  let result: any = {}
```
Comment (hannoeru) [inline src/parseSfc.ts:34]: Use `Record<string, any>` type

File: src/parseSfc.ts [hunk 6]
```diff
@@ -26,25 +27,47 @@ export interface FileError extends Error {
...
+    try {
+      result = YAML.parse(block.content)
+    }
+    catch (e) {
+      err = e
+      valid = false
     }
   }
+
+  if (valid) return result
```
Comment (hannoeru) [inline src/parseSfc.ts:64]: Don't need to check valid, if this.error() emit it will stop at that point, so just `return result`

**Comment by JohnCampionJr** [comment]
Anything else, let me know :)

File: src/parseSfc.ts [hunk 7]
```diff
@@ -26,25 +27,47 @@ export interface FileError extends Error {
...
-      throw wrapped
+  else if (lang === 'yaml' || (lang === 'any' && !valid)) {
+    try {
+      result = YAML.parse(block.content)
+    }
+    catch (e) {
+      err = e
+      valid = false
     }
   }
```
Comment (hannoeru) [inline src/parseSfc.ts]: Set default lang to json5 in L31 and change code to:
```ts
if (json) {
  parse json
} elseif (yaml) {
  parse yaml
} elseif (json5) {
  parse json5
} else {
  error
}
```

Comment (JohnCampionJr) [reply src/parseSfc.ts]: I thought about that, but wanted it to be able to parse YAML without having to specify ```lang="yaml"``` everywhere

File: src/parseSfc.ts (hunk 4 above)
Comment (JohnCampionJr) [reply src/parseSfc.ts:66]: Right now, the error is only thrown once.

File: src/parseSfc.ts (hunk 7 above)
Comment (hannoeru) [reply src/parseSfc.ts]: I think default to Json5 is better. In this case you can add a options to define default parser.

Comment (JohnCampionJr) [reply src/parseSfc.ts]: OK, I'll add the option and change this.  That's a good solution.  OK if I make "any" an option?  Then we're both happy? :)

**Comment by JohnCampionJr** [comment]
All set!

**Comment by hannoeru** [comment]
Great, thanks!

**Pull request merged**